chain/index/events.go

```diff
@@ -0,0 +1,583 @@
+				}
+
+				ce.TipSetKey = ts.Key()
+
+				ce.MsgCid, err = cid.Cast(row.messageCid)
+				if err != nil {
+					return nil, xerrors.Errorf("parse message cid: %w", err)
+				}
+			}
+
+			ce.Entries = append(ce.Entries, types.EventEntry{
+				Flags: row.flags[0],
+				Key:   row.key,
+				Codec: row.codec,
+				Value: row.value,
+			})
+		}
+
+		if ce != nil {
+			ces = append(ces, ce)
+		}
+
+		if len(ces) == 0 {
+			return nil, nil
+		}
+
+		// collected event list is in inverted order since we selected only the most recent events
+		// sort it into height order
+		sort.Slice(ces, func(i, j int) bool { return ces[i].Height < ces[j].Height })
+
+		return ces, nil
+	}
+
+	values, query, err := makePrefillFilterQuery(f)
+	if err != nil {
+		return nil, xerrors.Errorf("failed to make prefill filter query: %w", err)
+	}
+
+	stmt, err := si.db.Prepare(query)
+	if err != nil {
+		return nil, xerrors.Errorf("prepare prefill query: %w", err)
+	}
+	defer func() { _ = stmt.Close() }()
+
+	ces, err := getEventsFnc(stmt, values)
+	if err != nil {
+		return nil, xerrors.Errorf("failed to get events: %w", err)
+	}
+	if len(ces) == 0 {
+		if f.MaxHeight > 0 {
+			head := si.cs.GetHeaviestTipSet()
+			if head == nil {
+				return nil, errors.New("failed to get head: head is nil")
+			}
+			headHeight := head.Height()
+			maxLookBackHeight := headHeight - maxLookBackForWait
+
+			if f.MaxHeight <= maxLookBackHeight {
+				return nil, si.checkTipsetIndexedStatus(ctx, f)
+			}
+		}
+
+		// there's no matching events for the filter, wait till index has caught up to the head and then retry
+		if err := si.waitTillHeadIndexed(ctx); err != nil {
+			return nil, xerrors.Errorf("failed to wait for head to be indexed: %w", err)
+		}
+		ces, err = getEventsFnc(stmt, values)
+		if err != nil {
+			return nil, xerrors.Errorf("failed to get events: %w", err)
+		}
+
+		if len(ces) == 0 {
+			return nil, si.checkTipsetIndexedStatus(ctx, f)
+		}
+	}
+
+	return ces, nil
+}
+
+func makePrefillFilterQuery(f *EventFilter) ([]any, string, error) {
+	clauses := []string{}
+	values := []any{}
+	joins := []string{}
+
+	if f.TipsetCid != cid.Undef {
+		clauses = append(clauses, "tm.tipset_key_cid=?")
+		values = append(values, f.TipsetCid.Bytes())
+	} else {
+		if f.MinHeight >= 0 && f.MinHeight == f.MaxHeight {
+			clauses = append(clauses, "tm.height=?")
+			values = append(values, f.MinHeight)
+		} else {
+			if f.MaxHeight >= 0 && f.MinHeight >= 0 {
+				clauses = append(clauses, "tm.height BETWEEN ? AND ?")
+				values = append(values, f.MinHeight, f.MaxHeight)
+			} else if f.MinHeight >= 0 {
+				clauses = append(clauses, "tm.height >= ?")
+				values = append(values, f.MinHeight)
+			} else if f.MaxHeight >= 0 {
+				clauses = append(clauses, "tm.height <= ?")
+				values = append(values, f.MaxHeight)
+			} else {
+				return nil, "", xerrors.Errorf("filter must specify either a tipset or a height range")
+			}
+		}
+	}
+
+	if f.TipsetCid == cid.Undef {
```

Comment:
```suggestion

	// unless asking for a specific tipset, we never want to see reverted historical events
```

Reply:
This is done.

Reply:
@aarshkshah1992 my main point here was that the `if` is unnecessary, just remove it and the `}` before it because that's what the previous block is doing and it can go in there. The comment is useful but the `if` is unnecessary.

Reply:
Done.

Reply:
not sure why I didn't do this previously for the reverted bug .. hopefully we're not overlooking something!